### ISSUE TITLE
build issue on alpine linux in tinyxml2 (fseeko64/fseeko) embedded in gpds-1.8.1 (solution: optional add use system package version?)

### ISSUE THREAD

**Issue by bjornstromberg** [simulton/gpds#18]
the build is run on QSchematic 1.5.1 with gpds via FetchContent, the old qschematic-{1.4.0,1.5.0}/gdps 1.7.0 i've built separately on Alpine Linux as system packages without issues before.

just thought you should know its a identified corner case, and the alpine v3.19 tinyxml2 package is patched with https://git.alpinelinux.org/aports/tree/main/tinyxml2/no-lfs64.patch?h=3.19-stable this patch did not exist in alpine v3.18

so the real ask is to make it possible to get the 3rd party dependencies via find_package / find_package(pkgconf) instead right now they are embedded into gdps which makes it more work to patch downstream for consumers of the library.

ofcourse i can patch this privately downstream to suit my preferences, but it comes more people to use if its done upstream.

```
/home/user/build-dir/QSchematic> /usr/bin/ninja install
[5/61] Building CXX object _deps/gpds-build/lib/CMakeFiles/gpds-shared.dir/src/archiver_xml.cpp.o
[6/61] Building CXX object _deps/gpds-build/lib/CMakeFiles/gpds-static.dir/3rdparty/miniyaml/yaml/yaml.cpp.o
/home/user/build-dir/QSchematic/_deps/gpds-src/lib/3rdparty/tinyxml2/tinyxml2.cpp:2335:5: error: use of undeclared identifier 'fseeko64'; did you mean 'fseeko'?
 2335 |     TIXML_FSEEK( fp, 0, SEEK_SET );
      |     ^~~~~~~~~~~
      |     fseeko
/home/user/build-dir/QSchematic/_deps/gpds-src/lib/3rdparty/tinyxml2/tinyxml2.cpp:110:22: note: expanded from macro 'TIXML_FSEEK'
  110 |         #define TIXML_FSEEK fseeko64
      |                             ^
/usr/include/stdio.h:143:5: note: 'fseeko' declared here
  143 | int fseeko(FILE *, off_t, int);
      |     ^
/home/user/build-dir/QSchematic/_deps/gpds-src/lib/3rdparty/tinyxml2/tinyxml2.cpp:2341:5: error: use of undeclared identifier 'fseeko64'; did you mean 'fseeko'?
 2341 |     TIXML_FSEEK( fp, 0, SEEK_END );
      |     ^~~~~~~~~~~
      |     fseeko
/home/user/build-dir/QSchematic/_deps/gpds-src/lib/3rdparty/tinyxml2/tinyxml2.cpp:110:22: note: expanded from macro 'TIXML_FSEEK'
  110 |         #define TIXML_FSEEK fseeko64
      |                             ^
/usr/include/stdio.h:143:5: note: 'fseeko' declared here
  143 | int fseeko(FILE *, off_t, int);
      |     ^
/home/user/build-dir/QSchematic/_deps/gpds-src/lib/3rdparty/tinyxml2/tinyxml2.cpp:2345:44: error: use of undeclared identifier 'ftello64'; did you mean 'ftello'?
 2345 |         const long long fileLengthSigned = TIXML_FTELL( fp );
      |                                            ^~~~~~~~~~~
      |                                            ftello
/home/user/build-dir/QSchematic/_deps/gpds-src/lib/3rdparty/tinyxml2/tinyxml2.cpp:111:22: note: expanded from macro 'TIXML_FTELL'
  111 |         #define TIXML_FTELL ftello64
      |                             ^
/usr/include/stdio.h:144:7: note: 'ftello' declared here
  144 | off_t ftello(FILE *);
      |       ^
/home/user/build-dir/QSchematic/_deps/gpds-src/lib/3rdparty/tinyxml2/tinyxml2.cpp:2346:9: error: use of undeclared identifier 'fseeko64'; did you mean 'fseeko'?
 2346 |         TIXML_FSEEK( fp, 0, SEEK_SET );
      |         ^~~~~~~~~~~
      |         fseeko
/home/user/build-dir/QSchematic/_deps/gpds-src/lib/3rdparty/tinyxml2/tinyxml2.cpp:110:22: note: expanded from macro 'TIXML_FSEEK'
  110 |         #define TIXML_FSEEK fseeko64
      |                             ^
/usr/include/stdio.h:143:5: note: 'fseeko' declared here
  143 | int fseeko(FILE *, off_t, int);
      |     ^
4 errors generated.
ninja: job failed: /usr/bin/clang++ -Dgpds_shared_EXPORTS -I/home/user/build-dir/QSchematic/_deps/gpds-src/lib/include/gpds -O2 -g -DNDEBUG -std=gnu++20 -fPIC -MD -MT _deps/gpds-build/lib/CMakeFiles/gpds-shared.dir/3rdparty/tinyxml2/tinyxml2.cpp.o -MF _deps/gpds-build/lib/CMakeFiles/gpds-shared.dir/3rdparty/tinyxml2/tinyxml2.cpp.o.d -o _deps/gpds-build/lib/CMakeFiles/gpds-shared.dir/3rdparty/tinyxml2/tinyxml2.cpp.o -c /home/user/build-dir/QSchematic/_deps/gpds-src/lib/3rdparty/tinyxml2/tinyxml2.cpp
ninja: subcommand failed
*** Failure: Exit code 1 ***
```

**Comment by Tectu** [comment]
Allowing GPDS to consume `tinyxml2` via `find_package()` is a very sane proposal.

Would a mechanism similar to how `QSchematic` consumes `GPDS` be suitable in your opinion? i.e. there would be a CMake option like `GPDS_DEPENDENCY_TINYXML2_FROM_SYSTEM` or similar.

**Comment by Tectu** [comment]
Have a look at fc73f3357a40d74e613e9cefab0fe8739319f4a1 and let me know whether that works out for you.
I decided to retain the current default behavior (i.e. you'd have to explicitly set the CMake variable `GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE` to `ON`).

I can pump out a patch release if that helps your packaging/porting efforts.

**Comment by bjornstromberg** [comment]
now that was a fast response..  (ps. `I can get this done over the weekend™` is probably something i should hang on the wall as a big poster.. :roll_eyes: )

I'll have a go at trying to package it with just find_package, but i think it needs to try first find package, if that fails fallback to using cmake pkgconf to find tinyxml2 since downstream alot of code that supports cmake are still built with deploying old pkgconf (.pc files)

it seems that is the way on alpine currently [contents of tinyxml2-dev v3.18](https://pkgs.alpinelinux.org/contents?branch=v3.18&name=tinyxml2-dev&arch=x86_64&repo=main) , [contents of tinyxml2-dev v3.19](https://pkgs.alpinelinux.org/contents?branch=v3.19&name=tinyxml2-dev&arch=x86_64&repo=main) and [contents of tinyxml2-dev edge](https://pkgs.alpinelinux.org/contents?branch=v3.19&name=tinyxml2-dev&arch=x86_64&repo=main)

so unless I'm still to tired from not having my morning coffee yet, that will only try to use find_package, and not the find_package(PkgConf) train of calls (which would give a link name of PkgConf::tinyxml2


while I'm at it, do you want the resulting APKBUILD for gpds/qschematic I've made for my usage,
since there is no open source end application that uses it i don't think alpine will host the packages as of current.

the likely only maintenence for them is to update the version number and mabye run a ci with docker alpine containers if thorough..

/Björn

**Comment by bjornstromberg** [comment]
the gpds-config.cmake should probably notice that system version is used and use find_dependency either via tinyxml2::tinyxml2 or PkgConfig::tinyxml2

anyway got a little slice of time to look at this now, so will try and get it to play nice on-top  of fc73f3357a40d74e613e9cefab0fe8739319f4a1

**Comment by bjornstromberg** [comment]
quicker then a fork -> MR -> etc.. but github does not allow to upload patches so workaround like this..

`0002-patch-tinyxml-via-pkgconf.patch`

```
commit 3fd371880ab5a413d20c3047853ee2fcb0e765f1
Author: Björn Strömberg <bjorn.stromberg86@gmail.com>
Date:   Thu Feb 1 07:49:35 2024 +0100

    make use of PkgConfig as fallback incase find_package(tinyxml2) fails

diff --git a/lib/CMakeLists.txt b/lib/CMakeLists.txt
index 0249b58..c6b0953 100644
--- a/lib/CMakeLists.txt
+++ b/lib/CMakeLists.txt
@@ -29,9 +29,18 @@ set(HEADERS_PUBLIC
 # Dependency (tinyxml2)
 if (GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE)
 	find_package(
-		tinyxml2
-		REQUIRED
+		tinyxml2 QUIET
 	)
+
+	set(TINYXML2_TARGET tinyxml2::tinyxml2)
+
+	if(NOT tinyxml2_FOUND)
+		find_package(PkgConfig REQUIRED)
+		pkg_check_modules(tinyxml2 REQUIRED IMPORTED_TARGET tinyxml2)
+
+		set(TINYXML2_TARGET PkgConfig::tinyxml2)
+	endif()
+
 else()
 	list(APPEND HEADERS_PRIVATE
 		3rdparty/tinyxml2/tinyxml2.h
@@ -79,7 +88,7 @@ function(setup_target_common target)
 	target_link_libraries(
 		${target}
 		PRIVATE
-			$<$<BOOL:${GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE}>:tinyxml2::tinyxml2>
+			$<$<BOOL:${GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE}>:${TINYXML2_TARGET}>
 	)
 
 	set_target_properties(

```

**Comment by bjornstromberg** [comment]
will have to check on the possible linking issues with system tinyxml in a couple of hours some errands to do now.

but that one makes it compile on alpine linux via cmake using pkgconf

**Comment by bjornstromberg** [comment]
ps as expected after the packaging of the patched gpds using pkgconf

```
CMake Error at /usr/lib/cmake/gpds/gpds-targets.cmake:70 (set_target_properties):
  The link interface of target "gpds::gpds-static" contains:

    PkgConfig::tinyxml2

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  /usr/lib/cmake/gpds/gpds-config.cmake:1 (include)
  qschematic/external.cmake:25 (find_package)
  qschematic/CMakeLists.txt:2 (include)

```

will fix that when i get back in.

**Comment by bjornstromberg** [comment]
`0003-patch-export-tinyxml2-dependency.patch`
```
commit fbc8e846ed10f28a1f016526bca7dc95987b2b97
Author: Björn Strömberg <bjorn.stromberg86@gmail.com>
Date:   Thu Feb 1 08:20:53 2024 +0100

    export the tinyxml2 dependency

diff --git a/lib/CMakeLists.txt b/lib/CMakeLists.txt
index c6b0953..e288769 100644
--- a/lib/CMakeLists.txt
+++ b/lib/CMakeLists.txt
@@ -33,12 +33,14 @@ if (GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE)
 	)
 
 	set(TINYXML2_TARGET tinyxml2::tinyxml2)
+	set(GPDS_DEPENDENCY_TINYXML2_PKGCONF OFF)
 
 	if(NOT tinyxml2_FOUND)
 		find_package(PkgConfig REQUIRED)
 		pkg_check_modules(tinyxml2 REQUIRED IMPORTED_TARGET tinyxml2)
 
 		set(TINYXML2_TARGET PkgConfig::tinyxml2)
+		set(GPDS_DEPENDENCY_TINYXML2_PKGCONF ON)
 	endif()
 
 else()
diff --git a/lib/gpds-config.cmake b/lib/gpds-config.cmake
index 18302a2..d95fcb2 100644
--- a/lib/gpds-config.cmake
+++ b/lib/gpds-config.cmake
@@ -1 +1,14 @@
+
+set(GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE @GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE@)
+set(GPDS_DEPENDENCY_TINYXML2_PKGCONF @GPDS_DEPENDENCY_TINYXML2_PKGCONF@)
+
+if(GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE AND NOT GPDS_DEPENDENCY_TINYXML2_PKGCONF)
+  include(CMakeFindDependencyMacro)
+  find_dependency(tinyxml2 REQUIRED)
+else(GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE AND GPDS_DEPENDENCY_TINYXML2_PKGCONF)
+  include(CMakeFindDependencyMacro)
+  find_dependency(PkgConfig REQUIRED)
+  pkg_check_modules(tinyxml2 REQUIRED IMPORTED_TARGET tinyxml2)
+endif()
+
 include("${CMAKE_CURRENT_LIST_DIR}/gpds-targets.cmake")
```

it works, it leaves some cleaning up on variable names if you like the PKGCONF one to be an internal flag or something, but for now this compiles and packages clean on alpine.

**Comment by bjornstromberg** [comment]
whats needed now to finalize the APKBUILD script in case you want it added to the repo is to add those two patches on top of master, (my patch 0001- is fc73f3357a40d74e613e9cefab0fe8739319f4a1 extracted as a patch ontop of 1.8.1 tag, to cleanly 'package' the 1.8.1 release.

these patches would be nice to have a clean 1.8.2 tag for the final APKBUILD, so users only have to use a single file and no automatic extra patching that might break later, when they are part of a tag.  

but for now it works fine for me with qt6-qschematic-1.5.1 + gpds-1.8.1 with the patches in this thread.

**Comment by bjornstromberg** [comment]
if you rather want a MR, i can fix that too, but it seems overkill for this quick fix..

**Comment by Tectu** [comment]
> now that was a fast response.. (ps. `I can get this done over the weekend™` is probably something i should hang on the wall as a big poster.. 🙄 )

Yeah I edited that out once I figured that I'd just rather get it done now :D



> I'll have a go at trying to package it with just find_package, but i think it needs to try first find package, if that fails fallback to using cmake pkgconf to find tinyxml2 since downstream alot of code that supports cmake are still built with deploying old pkgconf (.pc files)
> 
> it seems that is the way on alpine currently [contents of tinyxml2-dev v3.18](https://pkgs.alpinelinux.org/contents?branch=v3.18&name=tinyxml2-dev&arch=x86_64&repo=main) , [contents of tinyxml2-dev v3.19](https://pkgs.alpinelinux.org/contents?branch=v3.19&name=tinyxml2-dev&arch=x86_64&repo=main) and [contents of tinyxml2-dev edge](https://pkgs.alpinelinux.org/contents?branch=v3.19&name=tinyxml2-dev&arch=x86_64&repo=main)

Is only shipping a `pkg-config` but not the CMake files provided by upstream (`tinyxml2` in this case) an alpine design decision?



> so unless I'm still to tired from not having my morning coffee yet, that will only try to use find_package, and not the find_package(PkgConf) train of calls (which would give a link name of PkgConf::tinyxml2
> 
> while I'm at it, do you want the resulting APKBUILD for gpds/qschematic I've made for my usage, since there is no open source end application that uses it i don't think alpine will host the packages as of current.
> 
> the likely only maintenence for them is to update the version number and mabye run a ci with docker alpine containers if thorough..

I cannot speak for alpine but plenty of Linux distros / OS seem to be quite happy to receive package contributions with no consumers yet as those packages can still be used by developers for their own projects that they might not publish or package. Furthermore, once there are consumers a fair bit of work is already done.
For example, I have [devel/qschematic](https://www.freshports.org/devel/qschematic) in the FreeBSD ports repo. It does not have a public consumer (yet) but it makes working on my currently non-public software projects easier.
Obviously, this is something that the alpine folks have to tell you - I'm just sharing experience here :)

As of right now, I don't think that I would necessarily want to put the `APKBUILD` into this repo as that would essentially mean that we as upstream have to maintain it.
Maybe you could publish it yourself elsewhere?




> 0003-patch-export-tinyxml2-dependency.patch

Is it possible that you're missing a call to `configure_file()`?
Also, we should probably rename `gpds-config.cmake` to `gpds-config.cmake.in` accordingly.


> if you rather want a MR, i can fix that too, but it seems overkill for this quick fix..

I'm perfectly fine with the `git-format` patches that you provide. Personally, I don't perceive any added value with the web GUI stuff that GitHub offers.



> these patches would be nice to have a clean 1.8.2 tag for the final APKBUILD, so users only have to use a single file and no automatic extra patching that might break later, when they are part of a tag.

Yep, I'll release a new patch version once this is done :)

**Comment by bjornstromberg** [comment]
> > I'll have a go at trying to package it with just find_package, but i think it needs to try first find package, if that fails fallback to using cmake pkgconf to find tinyxml2 since downstream alot of code that supports cmake are still built with deploying old pkgconf (.pc files)
> > it seems that is the way on alpine currently [contents of tinyxml2-dev v3.18](https://pkgs.alpinelinux.org/contents?branch=v3.18&name=tinyxml2-dev&arch=x86_64&repo=main) , [contents of tinyxml2-dev v3.19](https://pkgs.alpinelinux.org/contents?branch=v3.19&name=tinyxml2-dev&arch=x86_64&repo=main) and [contents of tinyxml2-dev edge](https://pkgs.alpinelinux.org/contents?branch=v3.19&name=tinyxml2-dev&arch=x86_64&repo=main)
> 
> Is only shipping a `pkg-config` but not the CMake files provided by upstream (`tinyxml2` in this case) an alpine design decision?

not really sure, I've asked for changes and gotten 'cmake' configs through but its always on edge (takes 6months until it gets to the newest version, and the old ones are unlikely. 

i target v3.18, but using 3.19 currently on dev pc due to their community repo with `the good stuff™` like KDE is located there only have support for 6months. the main repo has 2yr life cycle, which is why i target that. and build the rest myself, which is like chasing a moving bus..

> > so unless I'm still to tired from not having my morning coffee yet, that will only try to use find_package, and not the find_package(PkgConf) train of calls (which would give a link name of PkgConf::tinyxml2
> > while I'm at it, do you want the resulting APKBUILD for gpds/qschematic I've made for my usage, since there is no open source end application that uses it i don't think alpine will host the packages as of current.
> > the likely only maintenence for them is to update the version number and mabye run a ci with docker alpine containers if thorough..
> 
> I cannot speak for alpine but plenty of Linux distros / OS seem to be quite happy to receive package contributions with no consumers yet as those packages can still be used by developers for their own projects that they might not publish or package. Furthermore, once there are consumers a fair bit of work is already done. For example, I have [devel/qschematic](https://www.freshports.org/devel/qschematic) in the FreeBSD ports repo. It does not have a public consumer (yet) but it makes working on my currently non-public software projects easier. Obviously, this is something that the alpine folks have to tell you - I'm just sharing experience here :)

of course they like more packages.. if you maintain it.. which is where the problem is.. time.. i can maintain it for my own private usage but more then that i only have 24hours a day.. :roll_eyes: 

> As of right now, I don't think that I would necessarily want to put the `APKBUILD` into this repo as that would essentially mean that we as upstream have to maintain it. Maybe you could publish it yourself elsewhere?
> 

i'll throw up a repo on github so people can fetch it if they want it. reinventing the wheel ...

> > 0003-patch-export-tinyxml2-dependency.patch
> 
> Is it possible that you're missing a call to `configure_file()`? Also, we should probably rename `gpds-config.cmake` to `gpds-config.cmake.in` accordingly.
> 

yep it sure looks like it! good catch.. will fix that

> > if you rather want a MR, i can fix that too, but it seems overkill for this quick fix..
> 
> I'm perfectly fine with the `git-format` patches that you provide. Personally, I don't perceive any added value with the web GUI stuff that GitHub offers.
> 

totally agree.

> > these patches would be nice to have a clean 1.8.2 tag for the final APKBUILD, so users only have to use a single file and no automatic extra patching that might break later, when they are part of a tag.
> 
> Yep, I'll release a new patch version once this is done :)

great!

**Comment by bjornstromberg** [comment]
`0004-fix-configure_file-write-variables.patch`

```
commit ce249763f699afebb371259457a5559748096153
Author: Björn Strömberg <bjorn.stromberg86@gmail.com>
Date:   Thu Feb 1 16:13:05 2024 +0100

     #18: fix configure_file to actually write the variables like it should and not just copy. like Joel B pointed out in codereview.

diff --git a/lib/CMakeLists.txt b/lib/CMakeLists.txt
index e288769..42525d0 100644
--- a/lib/CMakeLists.txt
+++ b/lib/CMakeLists.txt
@@ -171,15 +171,15 @@ write_basic_package_version_file(
 )
 install(
 	FILES
-		gpds-config.cmake
+		"${CMAKE_CURRENT_BINARY_DIR}/gpds/gpds-config.cmake"
 		"${CMAKE_CURRENT_BINARY_DIR}/gpds/gpds-config-version.cmake"
-		DESTINATION ${ConfigPackageLocation}
+	DESTINATION ${ConfigPackageLocation}
 )
 
 # Package configuration
-configure_file(gpds-config.cmake
+configure_file(gpds-config.cmake.in
 	"${CMAKE_CURRENT_BINARY_DIR}/gpds/gpds-config.cmake"
-	COPYONLY
+	@ONLY
 )
 
 # Package export targets
diff --git a/lib/gpds-config.cmake b/lib/gpds-config.cmake.in
similarity index 100%
rename from lib/gpds-config.cmake
rename to lib/gpds-config.cmake.in
```

patch no. 4 this morning i was thinking one or two and this is done..  good with a set of extra eyes..

**Comment by Tectu** [comment]
> of course they like more packages.. if you maintain it.. which is where the problem is.. time.. i can maintain it for my own private usage but more then that i only have 24hours a day.. 🙄

As a package maintainer myself, I know the pain. That is why I try to keep the upstream that I'm the author of to be as nice as possible which is also why I was immediately like "yeah, definitely need to be able to pull in `tinyxml2` from the system`.
So much package maintenance pain is directly related to upstream "issues" (both technical and political). It's sometimes ridiculous.



> i'll throw up a repo on github so people can fetch it if they want it. reinventing the wheel ...

That sounds lovely (I totally get your snarky side remark - it is indeed hilarious...).
We'll gladly link to your repo in our `readme` if you like.




> 0004-fix-configure_file-write-variables.patch

Are these "stacked" patches now? Would it be possible for you to provide a single one touching everything we discussed? At least it seems that `0004` is no longer modifying the contents of `gpds-config.cmake(.in)`.

**Comment by bjornstromberg** [comment]
> > of course they like more packages.. if you maintain it.. which is where the problem is.. time.. i can maintain it for my own private usage but more then that i only have 24hours a day.. 🙄
> 
> As a package maintainer myself, I know the pain. That is why I try to keep the upstream that I'm the author of to be as nice as possible which is also why I was immediately like "yeah, definitely need to be able to pull in `tinyxml2` from the system`. So much package maintenance pain is directly related to upstream "issues" (both technical and political). It's sometimes ridiculous.
> 
it was about there i understood that you understood the real problem, and was actually interested in real solutions, not a faqcade of producing code. minimal maintenence in the end system while still keeping the boat floating is the real 'work'.

> > i'll throw up a repo on github so people can fetch it if they want it. reinventing the wheel ...
> 
> That sounds lovely (I totally get your snarky side remark - it is indeed hilarious...). We'll gladly link to your repo in our `readme` if you like.
> 

:+1: 

> > 0004-fix-configure_file-write-variables.patch
> 
> Are these "stacked" patches now? Would it be possible for you to provide a single one touching everything we discussed? At least it seems that `0004` is no longer modifying the contents of `gpds-config.cmake(.in)`.

i'll get you a single patch with all the commits, the 0001-4 is so APKBUILD puts the commits in the correct order ontop of 1.8.1 tarball.

yours will be a single patch with 0002-4 to load intop of master since you already have 0001 there..

**Comment by bjornstromberg** [comment]
single patchfile to load ontop of master: `fix-tinyxml2-system.patch` it contains the whole train, makes it easier.

```
From 3fd371880ab5a413d20c3047853ee2fcb0e765f1 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bj=C3=B6rn=20Str=C3=B6mberg?= <bjorn.stromberg86@gmail.com>
Date: Thu, 1 Feb 2024 07:49:35 +0100
Subject: [PATCH 1/3] make use of PkgConfig as fallback incase
 find_package(tinyxml2) fails

---
 lib/CMakeLists.txt | 15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git a/lib/CMakeLists.txt b/lib/CMakeLists.txt
index 0249b58..c6b0953 100644
--- a/lib/CMakeLists.txt
+++ b/lib/CMakeLists.txt
@@ -29,9 +29,18 @@ set(HEADERS_PUBLIC
 # Dependency (tinyxml2)
 if (GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE)
 	find_package(
-		tinyxml2
-		REQUIRED
+		tinyxml2 QUIET
 	)
+
+	set(TINYXML2_TARGET tinyxml2::tinyxml2)
+
+	if(NOT tinyxml2_FOUND)
+		find_package(PkgConfig REQUIRED)
+		pkg_check_modules(tinyxml2 REQUIRED IMPORTED_TARGET tinyxml2)
+
+		set(TINYXML2_TARGET PkgConfig::tinyxml2)
+	endif()
+
 else()
 	list(APPEND HEADERS_PRIVATE
 		3rdparty/tinyxml2/tinyxml2.h
@@ -79,7 +88,7 @@ function(setup_target_common target)
 	target_link_libraries(
 		${target}
 		PRIVATE
-			$<$<BOOL:${GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE}>:tinyxml2::tinyxml2>
+			$<$<BOOL:${GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE}>:${TINYXML2_TARGET}>
 	)
 
 	set_target_properties(
-- 
2.43.0


From fbc8e846ed10f28a1f016526bca7dc95987b2b97 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bj=C3=B6rn=20Str=C3=B6mberg?= <bjorn.stromberg86@gmail.com>
Date: Thu, 1 Feb 2024 08:20:53 +0100
Subject: [PATCH 2/3] export the tinyxml2 dependency

---
 lib/CMakeLists.txt    |  2 ++
 lib/gpds-config.cmake | 13 +++++++++++++
 2 files changed, 15 insertions(+)

diff --git a/lib/CMakeLists.txt b/lib/CMakeLists.txt
index c6b0953..e288769 100644
--- a/lib/CMakeLists.txt
+++ b/lib/CMakeLists.txt
@@ -33,12 +33,14 @@ if (GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE)
 	)
 
 	set(TINYXML2_TARGET tinyxml2::tinyxml2)
+	set(GPDS_DEPENDENCY_TINYXML2_PKGCONF OFF)
 
 	if(NOT tinyxml2_FOUND)
 		find_package(PkgConfig REQUIRED)
 		pkg_check_modules(tinyxml2 REQUIRED IMPORTED_TARGET tinyxml2)
 
 		set(TINYXML2_TARGET PkgConfig::tinyxml2)
+		set(GPDS_DEPENDENCY_TINYXML2_PKGCONF ON)
 	endif()
 
 else()
diff --git a/lib/gpds-config.cmake b/lib/gpds-config.cmake
index 18302a2..d95fcb2 100644
--- a/lib/gpds-config.cmake
+++ b/lib/gpds-config.cmake
@@ -1 +1,14 @@
+
+set(GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE @GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE@)
+set(GPDS_DEPENDENCY_TINYXML2_PKGCONF @GPDS_DEPENDENCY_TINYXML2_PKGCONF@)
+
+if(GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE AND NOT GPDS_DEPENDENCY_TINYXML2_PKGCONF)
+  include(CMakeFindDependencyMacro)
+  find_dependency(tinyxml2 REQUIRED)
+else(GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE AND GPDS_DEPENDENCY_TINYXML2_PKGCONF)
+  include(CMakeFindDependencyMacro)
+  find_dependency(PkgConfig REQUIRED)
+  pkg_check_modules(tinyxml2 REQUIRED IMPORTED_TARGET tinyxml2)
+endif()
+
 include("${CMAKE_CURRENT_LIST_DIR}/gpds-targets.cmake")
-- 
2.43.0


From ce249763f699afebb371259457a5559748096153 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bj=C3=B6rn=20Str=C3=B6mberg?= <bjorn.stromberg86@gmail.com>
Date: Thu, 1 Feb 2024 16:13:05 +0100
Subject: [PATCH 3/3]  #18: fix configure_file to actually write the variables
 like it should and not just copy. like Joel B pointed out in codereview.

---
 lib/CMakeLists.txt                              | 8 ++++----
 lib/{gpds-config.cmake => gpds-config.cmake.in} | 0
 2 files changed, 4 insertions(+), 4 deletions(-)
 rename lib/{gpds-config.cmake => gpds-config.cmake.in} (100%)

diff --git a/lib/CMakeLists.txt b/lib/CMakeLists.txt
index e288769..42525d0 100644
--- a/lib/CMakeLists.txt
+++ b/lib/CMakeLists.txt
@@ -171,15 +171,15 @@ write_basic_package_version_file(
 )
 install(
 	FILES
-		gpds-config.cmake
+		"${CMAKE_CURRENT_BINARY_DIR}/gpds/gpds-config.cmake"
 		"${CMAKE_CURRENT_BINARY_DIR}/gpds/gpds-config-version.cmake"
-		DESTINATION ${ConfigPackageLocation}
+	DESTINATION ${ConfigPackageLocation}
 )
 
 # Package configuration
-configure_file(gpds-config.cmake
+configure_file(gpds-config.cmake.in
 	"${CMAKE_CURRENT_BINARY_DIR}/gpds/gpds-config.cmake"
-	COPYONLY
+	@ONLY
 )
 
 # Package export targets
diff --git a/lib/gpds-config.cmake b/lib/gpds-config.cmake.in
similarity index 100%
rename from lib/gpds-config.cmake
rename to lib/gpds-config.cmake.in
-- 
2.43.0


```

**Comment by bjornstromberg** [comment]
note to self `git format-patch -3 --stdout > ~/fix-tinyxml2-system.patch` used `git show HEAD` before, that did not work so well for multiple commits in same file (it missed the 0002 in the middle)

**Comment by Tectu** [comment]
I've been preparing to push this but I feel like the automatic fall-back to `pkg-config` is unnecessarily complex.
How about we just add `GPDS_DEPENDENCY_TINYXML2_PKGCONF` as a high-level CMake `option()` analogous to `GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE`?
This seems easier to maintain, less complex, less error prone when modifying stuff down the road and I fail to conceive a scenario where you would not be able to specify this explicitly (i.e. in case of your alpine situation you'd use `GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE` for the alpine repo version(s) where the upstream CMake files are included and you'd use `GPDS_DEPENDENCY_TINYXML2_PKGCONF` for the alpine repo version(s) where upstream CMake files are not included).

Does that make sense?

**Comment by bjornstromberg** [comment]
> I've been preparing to push this but I feel like the automatic fall-back to `pkg-config` is unnecessarily complex. How about we just add `GPDS_DEPENDENCY_TINYXML2_PKGCONF` as a high-level CMake `option()` analogous to `GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE`? This seems easier to maintain, less complex, less error prone when modifying stuff down the road and I fail to conceive a scenario where you would not be able to specify this explicitly (i.e. in case of your alpine situation you'd use `GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE` for the alpine repo version(s) where the upstream CMake files are included and you'd use `GPDS_DEPENDENCY_TINYXML2_PKGCONF` for the alpine repo version(s) where upstream CMake files are not included).
> 
> Does that make sense?

its the simple way, reason for the fallback was due to the fact that you seemed to shrug at the need to use the ancient pkgconf solution, but i needed it. 

I'll have a go at it, and squash all the commits into one, for easier viewing now that we start getting close to a acceptable solution

**Comment by Tectu** [comment]
I was trying to understand the underlying reason or motivation with the hopes of being able to make better decisions. I have no problems supporting this. I'd just rather keep it simple.

Alright, once I get your patch I'll process that ASAP and provide you with a new release tag.

**Comment by bjornstromberg** [comment]
should be ready within a halfhour depending on if i messed up anything in the process... 

edit: two generator expressions doh..  good thing i did not say 5min..

**Comment by bjornstromberg** [comment]
well alpine pkg did not like building with squashed commits, i've manually tested to apply it on a testbranch and the commit it produces looks good.

`fix-tinyxml2-system.patch`
```
From 83261e24284d397b23950234a5c1b211b9570e3b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bj=C3=B6rn=20Str=C3=B6mberg?= <bjorn.stromberg86@gmail.com>
Date: Thu, 1 Feb 2024 07:49:35 +0100
Subject: [PATCH] add support for tinyxml2 via system supplied package, using
 options GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE and
 GPDS_DEPENDENCY_TINYXML2_PKGCONF

---
 CMakeLists.txt           | 35 ++++++++++++++++++++++++-----------
 lib/CMakeLists.txt       | 24 ++++++++++++++----------
 lib/gpds-config.cmake    |  1 -
 lib/gpds-config.cmake.in | 14 ++++++++++++++
 4 files changed, 52 insertions(+), 22 deletions(-)
 delete mode 100644 lib/gpds-config.cmake
 create mode 100644 lib/gpds-config.cmake.in

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 10e942b..4e0fdf7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -12,12 +12,24 @@ set(OPTION_BUILD_SHARED_DEFAULT ON)
 if (MSVC)
 	set(OPTION_BUILD_SHARED_DEFAULT OFF)
 endif()
-option(GPDS_BUILD_STATIC  					"Build static library." ON)
-option(GPDS_BUILD_SHARED  					"Build shared library." ${OPTION_BUILD_SHARED_DEFAULT})
-option(GPDS_BUILD_TESTS   					"Build the unit tests." OFF)
-option(GPDS_BUILD_EXAMPLES					"Build the examples." OFF)
-option(GPDS_FEATURE_SPDLOG					"Enable spdlog sink feature." OFF)
+option(GPDS_BUILD_STATIC                                   "Build static library." ON)
+option(GPDS_BUILD_SHARED                                   "Build shared library." ${OPTION_BUILD_SHARED_DEFAULT})
+option(GPDS_BUILD_TESTS                                    "Build the unit tests." OFF)
+option(GPDS_BUILD_EXAMPLES                                 "Build the examples." OFF)
+option(GPDS_FEATURE_SPDLOG                                 "Enable spdlog sink feature." OFF)
-option(GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE "Whether to use find_package() for the tinyxml2 dependency" OFF)
+option(GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE                "Whether to use find_package() for the tinyxml2 dependency" OFF)
+option(GPDS_DEPENDENCY_TINYXML2_PKGCONF                    "Whether to use find_package(PkgConfig) for the tinyxml2 dependency" OFF)
+
+if(GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE AND GPDS_DEPENDENCY_TINYXML2_PKGCONF)
+	message(FATAL_ERROR "GPDS: choose one of GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE and GPDS_DEPENDENCY_TINYXML2_PKGCONF not both.")
+endif()
+
+if(GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE OR GPDS_DEPENDENCY_TINYXML2_PKGCONF)
+	message(STATUS "GPDS: using system supplied tinyxml2 dependency")
+	set(_SYSTEM_TINYXML2 ON)
+else()
+	set(_SYSTEM_TINYXML2 OFF)
+endif()
 
 # User settings
 set(CMAKE_DEBUG_POSTFIX d)
@@ -56,16 +68,17 @@ endif()
 message(STATUS "")
 message(STATUS "------------------------")
 message(STATUS "GPDS configuration:")
-message(STATUS "  version       : " ${PROJECT_VERSION})
+message(STATUS "  version           : " ${PROJECT_VERSION})
 message(STATUS "")
 message(STATUS "  build:")
-message(STATUS "    static      : " ${GPDS_BUILD_STATIC})
-message(STATUS "    shared      : " ${GPDS_BUILD_SHARED})
-message(STATUS "    examples    : " ${GPDS_BUILD_EXAMPLES})
-message(STATUS "    tests       : " ${GPDS_BUILD_TESTS})
+message(STATUS "    static          : " ${GPDS_BUILD_STATIC})
+message(STATUS "    shared          : " ${GPDS_BUILD_SHARED})
+message(STATUS "    examples        : " ${GPDS_BUILD_EXAMPLES})
+message(STATUS "    tests           : " ${GPDS_BUILD_TESTS})
+message(STATUS "    system tinyxml2 : " ${_SYSTEM_TINYXML2})
 message(STATUS "")
 message(STATUS "  misc:")
-message(STATUS "    spdlog sink : " ${GPDS_FEATURE_SPDLOG})
+message(STATUS "    spdlog sink     : " ${GPDS_FEATURE_SPDLOG})
 message(STATUS "------------------------")
 message(STATUS "")
 
diff --git a/lib/CMakeLists.txt b/lib/CMakeLists.txt
index 0249b58..51bbb4e 100644
--- a/lib/CMakeLists.txt
+++ b/lib/CMakeLists.txt
@@ -27,11 +27,15 @@ set(HEADERS_PUBLIC
 )
 
 # Dependency (tinyxml2)
-if (GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE)
-	find_package(
-		tinyxml2
-		REQUIRED
-	)
+if (GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE OR GPDS_DEPENDENCY_TINYXML2_PKGCONF)
+	if(GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE)
+		find_package(tinyxml2 REQUIRED)
+		set(TINYXML2_TARGET tinyxml2::tinyxml2)
+	else()
+		find_package(PkgConfig REQUIRED)
+		pkg_check_modules(tinyxml2 REQUIRED IMPORTED_TARGET tinyxml2)
+		set(TINYXML2_TARGET PkgConfig::tinyxml2)
+	endif()
 else()
 	list(APPEND HEADERS_PRIVATE
 		3rdparty/tinyxml2/tinyxml2.h
@@ -79,7 +83,7 @@ function(setup_target_common target)
 	target_link_libraries(
 		${target}
 		PRIVATE
-			$<$<BOOL:${GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE}>:tinyxml2::tinyxml2>
+			$<$<OR:$<BOOL:${GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE}>,$<BOOL:${GPDS_DEPENDENCY_TINYXML2_PKGCONF}>>:${TINYXML2_TARGET}>
 	)
 
 	set_target_properties(
@@ -160,15 +164,15 @@ write_basic_package_version_file(
 )
 install(
 	FILES
-		gpds-config.cmake
+		"${CMAKE_CURRENT_BINARY_DIR}/gpds/gpds-config.cmake"
 		"${CMAKE_CURRENT_BINARY_DIR}/gpds/gpds-config-version.cmake"
-		DESTINATION ${ConfigPackageLocation}
+	DESTINATION ${ConfigPackageLocation}
 )
 
 # Package configuration
-configure_file(gpds-config.cmake
+configure_file(gpds-config.cmake.in
 	"${CMAKE_CURRENT_BINARY_DIR}/gpds/gpds-config.cmake"
-	COPYONLY
+	@ONLY
 )
 
 # Package export targets
diff --git a/lib/gpds-config.cmake b/lib/gpds-config.cmake
deleted file mode 100644
index 18302a2..0000000
--- a/lib/gpds-config.cmake
+++ /dev/null
@@ -1 +0,0 @@
-include("${CMAKE_CURRENT_LIST_DIR}/gpds-targets.cmake")
diff --git a/lib/gpds-config.cmake.in b/lib/gpds-config.cmake.in
new file mode 100644
index 0000000..d95fcb2
--- /dev/null
+++ b/lib/gpds-config.cmake.in
@@ -0,0 +1,14 @@
+
+set(GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE @GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE@)
+set(GPDS_DEPENDENCY_TINYXML2_PKGCONF @GPDS_DEPENDENCY_TINYXML2_PKGCONF@)
+
+if(GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE AND NOT GPDS_DEPENDENCY_TINYXML2_PKGCONF)
+  include(CMakeFindDependencyMacro)
+  find_dependency(tinyxml2 REQUIRED)
+else(GPDS_DEPENDENCY_TINYXML2_FINDPACKAGE AND GPDS_DEPENDENCY_TINYXML2_PKGCONF)
+  include(CMakeFindDependencyMacro)
+  find_dependency(PkgConfig REQUIRED)
+  pkg_check_modules(tinyxml2 REQUIRED IMPORTED_TARGET tinyxml2)
+endif()
+
+include("${CMAKE_CURRENT_LIST_DIR}/gpds-targets.cmake")
-- 
2.43.0

```

edit: apparently manually editing patches has not checksums..  anyway now the tabs are gone from the option lines

**Comment by bjornstromberg** [comment]
looks like the tabs mess up the layout on options, its a cosmetic thing but irritating that i saw it now.. kdevelop is setup to use spaces instead of tabs but i guess the files already had tabs so it used tabs..

edit: we're good to go..

**Comment by bjornstromberg** [comment]
this is good for a tag 'n push

**Comment by Tectu** [comment]
Can you check the current `master` branch? It's late here so might have screwed up.

I did remove the extra logic on those two options and simplified `gpds-config.cmake.in`.

**Comment by bjornstromberg** [comment]
`message(STATUS "    system tinyxml2 : " ${_SYSTEM_TINYXML2})` needs to be removed too, or fixed to handle the two options 

other than that it looks good, i did duplicate the include due to the fact that it might be unneeded for any non-system tinyxml2 case, which should be the normal case. 

guess the include is neglible in processing time and wont hurt.

**Comment by bjornstromberg** [comment]
ps late yep, its 03.15 at CET timezone.. so i should been sleeping for hours already.. 

edit: oh your swiss then you should be on CET too..

**Comment by Tectu** [comment]
The `v1.8.2` release has landed - Thank you for your contribution!

**Comment by bjornstromberg** [comment]
yep its already building, soon done.. thanks
